keps/sig-api-machinery/storage-migration.md

```diff
@@ -81,158 +81,109 @@ As to the deliverables,
-* watches the **migration worker** jobs. Updates
-  `migration.status.completedGeneration` when a job completes.
-* short-circuit: if `migration.spec.groupVersions` shows that the resource has
-  only one version, then its `.status.observedGeneration` and
-  `.status.completedGeneration` are updated directly without launching the
-  worker `job`.
+* watches the instances of the `migration` CR, adds them to a [workqueue][].
+* syncs one `migration` at a time to avoid overloading the apiserver,
+  * if `migration.status.completed == false`, it creates a **migration worker**
+    `job` to migrate the resource type. The `ownerReference` of the `job` is set
+    to the `migration`.
+  * waits until the **migration worker** `job` finishes, updates
+    `migration.status.completed` to `true`, and deletes the `job`.
 
 The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
 --include=<resource type>` to migrate a resource type. The **migrator worker**
 uses API chunking to retrieve partial lists of a resource type and thus can
-migrate a small chunk at a time. It stores the [continue token] in a `configmap`
-to preserve progress. With the inconsistent continue token introduced in
-[#67284][], the **migrator worker** does not need to worry about expired continue
-token.
+migrate a small chunk at a time. It stores the [continue token] in the owner
+`migration.status.continueToken`. With the inconsistent continue token
+introduced in [#67284][], the **migrator worker** does not need to worry about
+expired continue token.
 
+[list]:https://github.com/openshift/origin/blob/2a8633598ef0dcfa4589d1e9e944447373ac00d7/pkg/oc/cli/admin/migrate/storage/storage.go#L120-L184
+[workqueue]:https://github.com/kubernetes/client-go/tree/master/util/workqueue
 [#67284]:https://github.com/kubernetes/kubernetes/pull/67284
-[API section]:#api
+[API section]:#alpha-api
+
+The cluster admin can run the `kubectl get migrations -o
+jsonpath="{.status.completed}"` to see if all migrations have completed.
 
-The cluster admin can run the `kube-migrator progress` command to check if the
-migration has completed. The command lists all `migrations` instances and checks
-if all of them have `.spec.generation == .status.observedGeneration`.
+We will support a `kube-migrator migrate --resources=<list of resources>`
+command. It (re)creates `migration` CRs for the resources to trigger their
+migration. This command facilitates manually migrating custom resources and
+aggregated resources.
 
-### API
+### Alpha API
 
-We introduce the `migration` API to record the intention and progress of a
+We introduce the `migration` API to record the intention and the progress of a
 migration. The API will be a CRD defined in the `migration.k8s.io` group.
 
 Read the [workflow section][] to understand how the API is used.
 
 ```golang
 type StorageVersionMigration struct {
   metav1.TypeMeta
+  // metadata.name should be "<resource>.<group>" of the resource being
```

Comment:
Why? I'd probably just say to use generateName to avoid collisions. Requiring this means the old request has to be deleted before I can make a new request.

Reply:
I guess you want to keep the old requests to record history?

I'll have to add a `.status.active bool` to the `migration`, to let clients tell if the `migration` is for history keeping or is active. Sounds good?

Reply:
No, you already have a .status.completed for that. It's not about keeping history (which I don't want to do) it's about there not being any need to (effectively) have a per-resource lock on creating these things. (which is not even that effective since occasionally we move things between groups...)

Also imagine I have foo in group/v1 and also foo in group/v5, but they don't share key spaces because the one in v5 is a complete rewrite. Now I want to migrate v1 to v2 and also v5 to v6. It's contrived but I don't see any benefit at all from requiring the name match that format.

Reply:
Since kubernetes doesn't support transactions, I think it's easiest to only keep one `migration` for each resource type. We can keep the historic requests in a subfield.

Reply:
If the format of the name is `hash(etcd instance, etcd path prefix)` (as the next line in the comment suggested for the future), there is no issue in the case where we move things between groups, or in the case where group/v1/foo and group/v5/foo don't share key spaces.

It doesn't make sense to run multiple migration for the same resource type at the same time. If we go with your suggestion, for each resource type, we need to mark only one of the `migration` as active. Because Kubernetes doesn't support transactions, I don't see any easy way to do that. So I still favor the current design for simplicity.